discourse: 2.7.5 -> 2.7.7, plugin updates and fixes

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
- Update `discourse` to the latest stable version
- Update the plugins
- Run tests on the correct package, i.e. `discourseAllPlugins.tests` now runs the tests on the `discourseAllPlugins` package
- Change the auto generated plugin assets path to one that is writable at asset build time (fixes the issue in #132475)
- Make the `update.py` script able to package plugins from scratch

Closes #132112
Closes #133949

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

talyz committed:
discourse: 2.7.5 -> 2.7.7

talyz committed:
discourse.plugins.discourse-calendar: Update

talyz committed:
discourse.plugins.discourse-canned-replies: Update

talyz committed:
discourse.plugins.discourse-solved: Update

talyz committed:
discourse.plugins.discourse-data-explorer: Update

talyz committed:
discourse: Remove leftover link to unused plugins directory

talyz committed:
discourse.tests: Test the appropriate discourse package

Perform the tests on the package that the `tests` attribute is a child
of, i.e. if `discourseAllPlugins.tests` is built, the tests will run
with the `discourseAllPlugins` package, not the `discourse` package as
previously.

talyz committed:
discourse: Change the path to the auto generated plugin assets

Change the path to the auto generated plugin assets, which defaults to
the plugin's directory and isn't writable at the time of asset
generation.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 133611` at 3b85a888 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>discourse</li>
    <li>discourseAllPlugins</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `replace`.
Near pkgs/servers/web-apps/discourse/default.nix:4:9:
```
  |
4 | , ruby, replace, gzip, gnutar, git, cacert, util-linux, gawk
  |         ^
```
</li>
    <li>warning: unused-argument

Unused argument: `meta`.
Near pkgs/servers/web-apps/discourse/default.nix:56:7:
```
   |
56 |     , meta ? null
   |       ^
```
</li>
    <li>warning: unused-argument

Unused argument: `src`.
Near pkgs/servers/web-apps/discourse/default.nix:58:7:
```
   |
58 |     , src
   |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

bundler is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/servers/web-apps/discourse/default.nix:224:5:
```
    |
224 |     buildInputs = [
    |     ^
```
</li>
  </ul>
</details>

ryantm commented on pkgs/servers/web-apps/discourse/plugins/discourse-calendar/Gemfile:
```diff
@@ -2,4 +2,7 @@
 
 source "https://rubygems.org"
 
+git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
```
@talyz Looks like maybe it is unnecessary with bundler 2: https://github.com/rubygems/bundler/issues/5476 Is there some other reason it is needed?

talyz replied:
It was added by bundler, so I don't know :shrug: In our case, it's probably fine to remove it, since we're fetching the dependencies ourselves anyway, but it will most likely just be added back the next time `update.py` is run.

SuperSandro2000 commented on nixos/tests/discourse.nix:
```diff
@@ -4,7 +4,7 @@
 
 import ./make-test-python.nix (
-  { pkgs, lib, ... }:
+  { pkgs, lib, pkg ? pkgs.discourse, ... }:
```
```suggestion
  { pkgs, lib, package ? pkgs.discourse, ... }:
```

SuperSandro2000 commented on nixos/tests/discourse.nix:
```diff
@@ -55,6 +55,7 @@ import ./make-test-python.nix (
 
         services.discourse = {
           enable = true;
+          package = pkg;
           inherit admin;
```
```suggestion
          inherit admin pkg;
```

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/default.nix:
```diff
@@ -294,7 +303,7 @@ let
       plugins = callPackage ./plugins/all-plugins.nix { inherit mkDiscoursePlugin; };
       ruby = rubyEnv.wrappedRuby;
-      tests = nixosTests.discourse;
+      tests = import ../../../../nixos/tests/discourse.nix { pkg = pkgs.discourse.override args; };
```
```suggestion
      tests = import ../../../../nixos/tests/discourse.nix { package = pkgs.discourse.override args; };
```

talyz committed:
discourse.plugins.discourse-github: Update

talyz committed:
discourse.plugins: Make the updater able to package plugins

Let the update.py script handle the initial, repetitive task of
packaging new plugins. With this in place, the plugin only needs to be
added to the list in `update-plugins` and most of the work will be
done automatically when the script is run. Metadata still needs to be
filled in manually and some packages may of course require additional
work/patching.

talyz committed:
discourse.mkDiscoursePlugin: Handle repos with `gems` directories

Some plugin repos already have a `gems` directory. This lets the
packager choose whether it should be kept and the nix packaged ruby
gems should be copied into it or if it should be removed in favor of
our ruby gems.

talyz committed:
discourse: update.py: Remove native platforms in plugin lock files..

...and add ruby.

talyz commented:
I've made the `update.py` script able to package the initial versions of plugins now. With this, all you have to do to bootstrap a new plugin is add it to the list in the `update-plugins` function. I've also added handling of existing `gems` directories in plugin repos and made the plugin platform removal cover the plugin lock files as well.

talyz commented:
The test failed on ofborg due to a request timing out, but it passes locally.
